core/src/main/scala/org/http4s/client/jdkhttpclient/JdkWebSocketClient.scala

```diff
@@ -0,0 +1,129 @@
+                    : CompletionStage[_] =
+                  handleReceive(WSFrame.Binary(ByteVector(data), last).asRight)
+                override def onPing(webSocket: JWebSocket, message: ByteBuffer)
+                    : CompletionStage[_] =
+                  handleReceive(WSFrame.Ping(ByteVector(message)).asRight)
+                override def onPong(webSocket: JWebSocket, message: ByteBuffer)
+                    : CompletionStage[_] =
+                  handleReceive(WSFrame.Pong(ByteVector(message)).asRight)
+                override def onError(webSocket: JWebSocket, error: Throwable): Unit = {
+                  handleReceive(error.asLeft); ()
+                }
+              }
+              webSocket <- fromCompletableFuture(
+                F.delay(wsBuilder.buildAsync(URI.create(uri.renderString), wsListener))
+              )
+              sendSem <- Semaphore[F](1L)
+            } yield (webSocket, queue, sendSem)
+          } {
+            case (webSocket, _, _) =>
+              for {
+                isOutputOpen <- F.delay(!webSocket.isOutputClosed)
+                closeOutput = fromCompletableFuture(
+                  F.delay(webSocket.sendClose(JWebSocket.NORMAL_CLOSURE, ""))
+                )
+                _ <- closeOutput.whenA(isOutputOpen)
+                _ <- F.delay(webSocket.abort())
+              } yield ()
+          }
+          .map {
+            case (webSocket, queue, sendSem) =>
+              // sending will throw if done in parallel
+              val rawSend = (wsf: WSFrame) =>
+                fromCompletableFuture(F.delay(wsf match {
+                  case WSFrame.Text(text, last) => webSocket.sendText(text, last)
+                  case WSFrame.Binary(data, last) => webSocket.sendBinary(data.toByteBuffer, last)
+                  case WSFrame.Ping(data) => webSocket.sendPing(data.toByteBuffer)
+                  case WSFrame.Pong(data) => webSocket.sendPong(data.toByteBuffer)
+                  case WSFrame.Close(statusCode, reason) => webSocket.sendClose(statusCode, reason)
+                })).void
+              new WSConnection[F] {
+                override def send(wsf: WSFrame) =
+                  sendSem.withPermit(rawSend(wsf))
+                override def sendMany[G[_]: Traverse, A <: WSFrame](wsfs: G[A]) =
+                  sendSem.withPermit(wsfs.traverse_(rawSend))
+                override def receive =
+                  F.delay(webSocket.request(1)) *> queue.dequeue1.map(_.sequence).rethrow
+                override def subprotocol =
+                  webSocket.getSubprotocol.some.filter(_.nonEmpty)
+              }
+          }
+    }
+
+  /** A `WebSocketClient` wrapping the default `HttpClient`. */
+  def simple[F[_]](implicit F: ConcurrentEffect[F]): F[WebSocketClient[F]] =
+    F.delay(HttpClient.newHttpClient()).map(apply(_))
+
+  private def toCompletionStage[F[_], A](fa: F[A])(implicit F: Effect[F]): CompletionStage[A] = {
```

Comment:
same story as for `fromCompletableFuture` - should this be an internal helper function in `http4s-core`?

Reply:
I want it in cats-effect, but I've been shot down.  Yes, http4s-core internals seems like an appropriate place, so other modules can use it.

Reply:
And probably call it unsafeToCompletionStage.

Reply:
I just revisited `fromFuture` last night, and am left reconsidering this: as `internals` are exposed in other repos in the org, and once we do that, we need to be careful about preserving binary compatibility.  This is one of the disadvantages to the polyrepo approach.  Maybe `org.http4s.util` is a legitimate package, though I almost always regret utils libraries.

Or maybe we just copy and paste it everywhere.  I don't know.